crates/core/executor/src/memory.rs

```diff
@@ -0,0 +1,290 @@
+use std::mem::{replace, size_of};
+
+use serde::{Deserialize, Serialize};
+use vec_map::VecMap;
+
+use crate::events::MemoryRecord;
+
+/// The base 2 logarithm of the (maximum) page size in bytes.
+pub const LOG_PAGE_SIZE: usize = 12;
+/// The base 2 logarithm of the length of each page, considered as an array of `Option<MemoryRecord>`.
+pub const LOG_PAGE_LEN: usize =
+    LOG_PAGE_SIZE - size_of::<Option<MemoryRecord>>().next_power_of_two().ilog2() as usize;
+/// The length of each page, considered as an array of `Option<MemoryRecord>`.
+pub const PAGE_LEN: usize = 1 << LOG_PAGE_LEN;
+/// The mask for retrieving the lowest bits necessary to index within a page.
+pub const PAGE_MASK: usize = PAGE_LEN - 1;
+/// The maximum number of pages. Used for the length of the page table.
+pub const MAX_PAGE_COUNT: usize = 1 << (u32::BITS as usize - LOG_PAGE_LEN - NUM_IGNORED_LOWER_BITS);
+/// The number of lower bits to ignore, since addresses (except registers) are a multiple of 4.
+pub const NUM_IGNORED_LOWER_BITS: usize = 2;
+/// The number of registers in the virtual machine.
+pub const NUM_REGISTERS: usize = 32;
+
+/// A page of memory.
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct Page<V>(VecMap<V>);
+
+impl<V> Page<V> {
+    /// Create a `Page` with capacity `PAGE_LEN`.
+    pub fn new() -> Self {
+        Self(VecMap::with_capacity(PAGE_LEN))
+    }
+}
+
+impl<V> Default for Page<V> {
+    fn default() -> Self {
+        Self(VecMap::default())
+    }
+}
+
+/// Paged memory. Balances both memory locality and total memory usage.
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct PagedMemory<V> {
+    /// The internal page table.
+    pub page_table: VecMap<Page<V>>,
```

Comment:
VecMap actually has [worst case O(n) lookup and performance](https://docs.rs/vecmap-rs/latest/vecmap/), it seems incorrect to use in both places here?

Reply:
This is the crate `vec_map`.